README.md
[![Build Status](https://travis-ci.org/janstuemmel/grav-plugin-srcset.svg?branch=master)](https://travis-ci.org/janstuemmel/grav-plugin-srcset)

# Srcset Plugin

The **Srcset** Plugin provides support for the srcset attribute in rendered markdown content. Instead of manipulating the rendered content, it extends the markdown parser. It wont't work with editors that are not using the Grav Parsedown Parser. So it will not work with tinymce editor plugin.  

## Manual Installation

To install the plugin manually, download the zip-version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `srcset`. You can find these files on [GitHub](https://github.com//grav-plugin-srcset) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/srcset

## Configuration

Before configuring this plugin, you should copy the `user/plugins/srcset/srcset.yaml` to `user/config/plugins/srcset.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
enabled: true
```

Note that if you use the Admin Plugin, a file with your configuration named srcset.yaml will be saved in the `user/config/plugins/`-folder once the configuration is saved in the Admin.


CHANGELOG.md
# v0.1.0
##  07/26/2019

1. [](#new)
    * ChangeLog started...


spec/functions.spec.php
<?php

use Spec\ImageMock;

describe('tests', function() {

  beforeEach(function() {
    $this->config = [ 
      'widths' => [
        [ 'image_width' => 480, 'inherent_width' => 480 ], 
        [ 'image_width' => 960, 'inherent_width' => 960 ] 
      ],
    ];
  });


  it('should set `srcset` and remove `src` attribute', function() {

    // given
    $images = [ 'bar.jpeg' => new ImageMock('/foo/bar.jpeg') ];
    $Inline = [
      'element' => [
        'attributes' => [
          'src' => 'bar.jpeg'
        ]
      ]
    ];

    // when
    $el = manipulateElement($Inline, $images, $this->config); 
    
    // then
    expect($el['element']['attributes']['srcset'])->toBe('/foo/bar.jpeg?mocked 480w,/foo/bar.jpeg?mocked 960w');
    expect(isset($el['element']['attributes']['src']))->toBeFalsy();
  });


  it('should not set `srcset` when image not in page media', function() {

    // given
    $Inline = [
      'element' => [
        'attributes' => [
          'src' => 'bar.jpeg'
        ]
      ]
    ];

    // when
    $el = manipulateElement($Inline, [], $this->config); 

    // // then
    expect($el['element']['attributes']['src'])->toBe('bar.jpeg');
    expect(isset($el['element']['attributes']['srcset']))->toBeFalsy();
  });


  it('should apply image manipulation arguments', function() {

    // given
    $image = new ImageMock('/foo/bar.jpeg');
    $images = [ 'bar.jpeg' => $image ];
    $Inline = [
      'element' => [
        'attributes' => [
          'src' => 'bar.jpeg?sepia&gaussianBlur=2'
        ]
      ]
    ];

    // then
    expect($image)->toReceive('sepia')->with('')->times(2);
    expect($image)->toReceive('gaussianBlur')->with('2')->times(2);
    expect($image)->toReceive('resize')->with(480)->with(960);
    // expect($image)->toReceive('resize')->times(1);

    // when
    $el = manipulateElement($Inline, $images, $this->config); 
  });
  

  it('should not increase image size when original image is smaller', function() {

    // given
    $image = new ImageMock('/foo/bar.jpeg', [ 500, 500 ]);
    $images = [ 'bar.jpeg' => $image ];
    $Inline = [
      'element' => [
        'attributes' => [
          'src' => 'bar.jpeg'
        ]
      ]
    ];   

    // then
    expect($image)->toReceive('resize')->times(1);
    expect($image)->toReceive('resize')->with(480);

    // when
    $el = manipulateElement($Inline, $images, $this->config); 
  });

});




srcset.php
<?php
namespace Grav\Plugin;

use Grav\Common\Plugin;
use Grav\Common\Grav;
use RocketTheme\Toolbox\Event\Event;

use \Grav\Common\Page\Medium\ImageMedium;

class SrcsetPlugin extends Plugin {

    public static function getSubscribedEvents() {
        return [
            'onPluginsInitialized' => ['onPluginsInitialized', 0]
        ];
    }

    public static function getImageMagicActions() {
        
        $actions = [];

        foreach(ImageMedium::$magic_actions as $action) {
            $actions[$action] = $action;
        }

        return $actions;
    }

    public function onPluginsInitialized() {

        if ($this->isAdmin()) {
            return;
        }

        $this->enable([
            'onMarkdownInitialized' => ['onMarkdownInitialized', 0]
        ]);
    }

    public function onMarkdownInitialized($e) {

        $page = $e['page'];
        $markdown = $e['markdown'];

        $images = $page->media()->images();

        $config = $this->config->get('plugins.srcset');

        $this->grav['assets']->addJs('http://afarkas.github.io/lazysizes/lazysizes.min.js');

        
        foreach($images as $image) {
            $this->grav['debugger']->addMessage($this->grav['page']);
            // $image->resize(500);
            // $image->derivatives([400,800,1000,1200])->sizes('80vw');
            // $image->derivatives([400,800,1000,1200]);
            // ->sizes('(min-width: 400px) 800px, 50vw');
        }

        $markdown->addInlineType('!', 'ImageExtended', 0);

        require_once __DIR__ . '/src/functions.php';

        $markdown->inlineImageExtended = inlineImageExtended($images, $config, $this->grav['debugger'])->bindTo($markdown, $markdown);        
                
    }
}


src/functions.php
<?php

function inlineImageExtended($images, $config = [], $debugger = null) {

  return function($Excerpt) use ($images, $config, $debugger) {

    $Inline = $this->inlineImage($Excerpt);
    $InlineOriginal = parent::inlineImage($Excerpt);

    $imageData = getOriginalImage($InlineOriginal['element']['attributes']['src'], $images);

    $Inline = manipulateElement($Inline, $imageData, $config, $debugger);
    
    return $Inline;
  };
}

function manipulateElement($Inline, $imageData, $config, $debuger = null) {
    
    if(!isset($Inline)) return null;
    
    if($imageData) {

      @[$image, $params] = $imageData;

      $image->derivatives([400,800,1000,1200]);
    }

    // return $Inline;
    // unset($Inline['element']['attributes']['sizes']);
    // $Inline['element']['attributes']['data-sizes'] = 'auto';



    $Inline['element']['attributes']['data-srcset'] = $Inline['element']['attributes']['srcset'];
    // $Inline['element']['attributes']['data-src'] = $Inline['element']['attributes']['src'];

    // $Inline['element']['attributes']['sizes'] = '(max-width: 400px) 800px, 90vw';
    // $Inline['element']['attributes']['data-sizes'] = '(max-width: 400px) 800px, 90vw';
    
    $Inline['element']['attributes']['class'] .= ' lazyload';

    unset($Inline['element']['attributes']['srcset']);
    unset($Inline['element']['attributes']['src']);

    return $Inline;

    // unset($Inline['element']['attributes']['sizes']);
    
    // set src fallback when image is not in page media
    if ($imageData) {
    }

    // set fallback when config fallback==true
    if ($imageData && $config['fallback']) {
      
      [$image, $params] = $imageData;

      $fallbackSettings = is_array($config['fallback_settings']) ? $config['fallback_settings'] : []; 

      foreach($fallbackSettings as $filter) {
        applyImageFilter($imgage, $filter['key'], $filter['value']);
      }

      $Inline['element']['attributes']['src'] = $image->url();
    }

    return $Inline;
}

// helper

function getOriginalImage($src, $images) {

  @['basename' => $basename, 'params' => $params] = parseImageSrc($src);
  
  if (!array_key_exists($basename, $images)) {
    return null;
  }

  return [ $images[$basename], $params];
}

function parseImageSrc($src) {
  
  // destruct array
  @['path' => $path, 'query' => $query] = parse_url($src);

  // parse query params
  parse_str($query, $params);

  return [
    'basename' => basename($path), 
    'params' => is_array($params) ? $params : [], 
  ];
}

function applyImageFilters($img, $params) {
  foreach ($params as $type => $value) {
    applyImageFilter($img, $type, $value);
  }
}

function applyImageFilter($img, $type, $value) {
  try {
    $img->__call($type, [ $value ]);
  } 
  // silently fail on invalid number of arguments or a filter doesnt exist
  catch (\InvalidArgumentException $e) {}
  catch (\BadFunctionCallException $e) {}
}
